packages/solana/src/providers/WalletConnectProvider.ts

```diff
@@ -0,0 +1,254 @@
+import UniversalProvider from '@walletconnect/universal-provider'
+import { SolConstantsUtil, type AnyTransaction, type Chain, type Provider } from '../utils/scaffold'
+import { ProviderEventEmitter } from './shared/ProviderEventEmitter'
+import type { SessionTypes } from '@walletconnect/types'
+import base58 from 'bs58'
+import {
+  Connection,
+  PublicKey,
+  Transaction,
+  VersionedTransaction,
+  type SendOptions
+} from '@solana/web3.js'
+import { isVersionedTransaction } from '@solana/wallet-adapter-base'
+
+export type WalletConnectProviderConfig = {
+  provider: UniversalProvider
+  chains: Chain[]
+}
+
+export class WalletConnectProvider extends ProviderEventEmitter implements Provider {
+  public readonly name = 'WalletConnect'
+  public readonly type = 'WALLET_CONNECT'
+  public readonly icon = 'https://avatars.githubusercontent.com/u/37784886'
+
+  private provider: UniversalProvider
+  private session?: SessionTypes.Struct
+  private readonly requestedChains: Chain[]
+
+  constructor({ provider, chains }: WalletConnectProviderConfig) {
+    super()
+    this.requestedChains = chains
+    this.provider = provider
+  }
+
+  // -- Universal Provider Events ------------------------ //
+  public onUri?: (uri: string) => void
+
+  // -- Public ------------------------------------------- //
+
+  public get chains() {
+    return (
+      (this.session?.namespaces['solana']?.chains
+        ?.map(sessionChainId => {
+          // This is a workaround for wallets that only accept Solana deprecated networks
+          let chainId = sessionChainId
+          if (chainId === SolConstantsUtil.CHAIN_IDS.Deprecated_Mainnet) {
+            chainId = SolConstantsUtil.CHAIN_IDS.Mainnet
+          } else if (chainId === SolConstantsUtil.CHAIN_IDS.Deprecated_Devnet) {
+            chainId = SolConstantsUtil.CHAIN_IDS.Devnet
+          }
+
+          return this.requestedChains.find(chain => `solana:${chain.chainId}` === chainId)
+        })
+        .filter(Boolean) as Chain[]) || []
+    )
+  }
+
+  public get publicKey() {
+    const account = this.getAccount(false)
+
+    if (account) {
+      return new PublicKey(account.publicKey)
+    }
+
+    return undefined
+  }
+
+  public async connect() {
+    const rpcMap = this.requestedChains.reduce<Record<string, string>>((acc, chain) => {
+      acc[`solana:${chain.chainId}`] = chain.rpcUrl
+
+      return acc
+    }, {})
+
+    if (this.provider.session) {
+      this.session = this.provider.session
+    } else {
+      this.provider.on('display_uri', this.onUri)
+      this.session = await this.provider.connect({
+        optionalNamespaces: {
+          solana: {
+            chains: this.getRequestedChainsWithDeprecated(),
+            methods: [
+              'solana_signMessage',
+              'solana_signTransaction',
+              'solana_signAndSendTransaction'
+            ],
+            events: [],
```

Comment:
Should we add the accountChanged event here?

Reply:
I have tried to add the event but then it doesn't connect, as main code also have this issue I decided to not tackle it in the refactor as may increase even more the PR size.